packages/react-reconciler/src/ReactFiberHotReloading.js

```diff
@@ -0,0 +1,258 @@
+        }
+        break;
+      }
+      case MemoComponent:
+      case SimpleMemoComponent: {
+        if (
+          typeof nextType === 'object' &&
+          nextType !== null &&
+          nextType.$$typeof === REACT_MEMO_TYPE
+        ) {
+          // TODO: if it was but can no longer be simple,
+          // we shouldn't set this.
+          needsCompareFamilies = true;
+        }
+        break;
+      }
+      // TODO: maybe support lazy?
+      default:
+        return false;
+    }
+
+    // Check if both types have a family and it's the same one.
+    if (needsCompareFamilies) {
+      // Note: memo() and forwardRef() we'll compare outer rather than inner type.
+      // This means both of them need to be registered to preserve state.
+      // If we unwrapped and compared the inner types for wrappers instead,
+      // then we would risk falsely saying two separate memo(Foo)
+      // calls are equivalent because they wrap the same Foo function.
+      const prevFamily = familiesByType.get(prevType);
+      if (
+        prevFamily !== undefined &&
+        prevFamily === familiesByType.get(nextType)
+      ) {
+        return true;
+      }
+    }
+    return false;
+  } else {
+    return false;
+  }
+}
+
+export function markFailedErrorBoundaryForHotReloading(fiber: Fiber) {
+  if (__DEV__) {
+    if (familiesByType === null) {
+      // Not hot reloading.
+      return;
+    }
+    if (typeof WeakSet !== 'function') {
+      return;
+    }
+    if (failedBoundaries === null) {
+      failedBoundaries = new WeakSet();
+    }
+    failedBoundaries.add(fiber);
+  }
+}
+
+export function scheduleHotUpdate(hotUpdate: HotUpdate): void {
+  if (__DEV__) {
+    // TODO: warn if its identity changes over time?
+    familiesByType = hotUpdate.familiesByType;
+
+    const {root, staleFamilies, updatedFamilies} = hotUpdate;
+    flushPassiveEffects();
+    flushSync(() => {
+      scheduleFibersWithFamiliesRecursively(
+        root.current,
+        updatedFamilies,
+        staleFamilies,
+      );
+    });
+    flushPassiveEffects();
+  }
+}
+
+function scheduleFibersWithFamiliesRecursively(
+  fiber: Fiber,
+  updatedFamilies: Set<Family>,
+  staleFamilies: Set<Family>,
+) {
+  if (__DEV__) {
+    const {alternate, child, sibling, tag, type} = fiber;
+
+    let candidateType = null;
+    switch (tag) {
+      case FunctionComponent:
+      case SimpleMemoComponent:
+        candidateType = type;
+        break;
+      case ForwardRef:
+        candidateType = type.render;
+        break;
+      default:
+        break;
+    }
+
+    if (familiesByType === null) {
+      throw new Error('Expected familiesByType to be set during hot reload.');
+    }
+
+    if (candidateType !== null) {
+      const family = familiesByType.get(candidateType);
+      if (family !== undefined) {
+        if (staleFamilies.has(family)) {
+          fiber._debugNeedsRemount = true;
+          scheduleWork(fiber, Sync);
```

Comment:
Same here. Do the hot updates need to be sync?

Reply:
They don't have to be but the developer is waiting and there's nothing happening that could be more important at that moment. It also makes it simpler to highlight updated nodes visually after the fact. (Unless I get notified about completion somehow.)